### PR TITLE
Handle unification of parameter and module reference types

### PR DESCRIPTION
**Pull request by Bargsteen**
## Purpose

Handle unification of the parameter and module reference types.

Depends on:
- https://github.com/Concordium/concordium-base/pull/333
- https://github.com/Concordium/concordium-contracts-common/pull/73
- https://github.com/Concordium/concordium-contracts-common/pull/75
- https://github.com/Concordium/concordium-base/pull/334

## Changes

- Use `OwnedParameter` and `ModuleReference` instead of the now deprecated `Parameter` and `ModuleRef`.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: examples/init-update-contract.rs [hunk 1]
```diff
@@ -8,11 +8,10 @@ use concordium_rust_sdk::{
     common::{types::TransactionTime, SerdeDeserialize, SerdeSerialize},
     endpoints,
     smart_contracts::common::{
-        self as contracts_common, Amount, ContractAddress, OwnedContractName, OwnedReceiveName,
-        Serial,
+        Amount, ContractAddress, OwnedContractName, OwnedReceiveName, Serial,
     },
     types::{
-        smart_contracts::{ModuleRef, Parameter},
+        smart_contracts::{ModuleReference, OwnedParameter},
```
Comment (abizjak) [inline examples/init-update-contract.rs:14]: Please document changes in the changelog please.

**Review by abizjak — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,8 @@
 - Add `find_earliest_finalized`, `find_instance_creation`,
   `find_first_finalized_block_no_later_than` methods to the `v2` client.
 - Bump MSRV to 1.60
+- Add deprecation notices to `ModuleRef` and `Parameter`. Use `ModuleReference`
```
[the remark this answers is not in the document]
Comment (Bargsteen) [reply CHANGELOG.md:6]: Yes, I thought it wasn't necessary as there are no breaking changes. But I've added information about additional methods and implementations.

Looking at `OwnedParameter::new`, I realised that it doesn't check the parameter size. I think we should change its type to return a `Result` instead. The impact should be minimal.

[the remark this answers is not in the document]
Comment (abizjak) [reply CHANGELOG.md:6]: As discussed in slack we'll reaplace ::new with from_serial which will be fallible

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,17 @@
...
+  and `OwnedParameter`, respectively, instead.
+  - There are no breaking changes to the API of either type, but `OwnedParameter` has some
```
Comment (abizjak) [inline CHANGELOG.md:8]: Isn't there a breaking change in that `Parameter` used to have `AsRef<Vec<u8>>` and now there is `AsRef<&[u8]>` only?

**Review by abizjak — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (Bargsteen) [reply CHANGELOG.md:8]: Ah yes, I forgot about that one. Thanks!

**Review by limemloh — APPROVED** [review]

**Pull request merged**